base/util/merge_test.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2020 Red Hat, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.)
+
+package util
+
+import (
+	"testing"
+
+	"github.com/coreos/fcct/translate"
+
+	"github.com/coreos/ignition/v2/config/util"
+	// config version doesn't matter; just pick one
+	"github.com/coreos/ignition/v2/config/v3_0/types"
+	"github.com/coreos/vcontext/path"
+	"github.com/stretchr/testify/assert"
+)
+
+// TestMergeTranslatedConfigs tests merging two Ignition configs and their
+// corresponding translations.
+func TestMergeTranslatedConfigs(t *testing.T) {
+	tests := []struct {
+		parent             types.Config
+		parentTranslations translate.TranslationSet
+		child              types.Config
+		childTranslations  translate.TranslationSet
+		merged             types.Config
+		mergedTranslations translate.TranslationSet
+	}{
+		{
+			parent: types.Config{
+				Ignition: types.Ignition{
+					Version: "3.0.0",
+				},
+				Systemd: types.Systemd{
+					Units: []types.Unit{
+						{
+							Name:     "aardvark.service",
+							Enabled:  util.BoolToPtr(true),
+							Contents: util.StrToPtr("antelope"),
+						},
+						{
+							Name:     "caribou.service",
+							Contents: util.StrToPtr("caribou"),
+						},
+						{
+							Name:     "elephant.service",
+							Contents: util.StrToPtr("elephant"),
+						},
+					},
+				},
+			},
+			parentTranslations: makeTranslationSet([]translate.Translation{
+				// parent key duplicated in child, should be clobbered
+				{path.New("in", "bad", 1), path.New("out", "systemd", "units", 0, "name")},
+				// parent field overridden in child, should be clobbered
+				{path.New("in", "bad", 2), path.New("out", "systemd", "units", 0, "contents")},
+				// parent field not overridden in child
+				{path.New("in", "good", 1), path.New("out", "systemd", "units", 0, "enabled")},
+				// parent key not specified in child
+				{path.New("in", "good", 2), path.New("out", "systemd", "units", 1, "name")},
+				// parent field not specified in child
+				{path.New("in", "good", 3), path.New("out", "systemd", "units", 1, "contents")},
+				// other fields omitted from translation set
```

Comment:
Is the idea behind omitting the `TranslationSet` for `elephant.service` to test that we can handle cases where there's a field in the result config / transcription that we don't have a corresponding `TranslationSet` for?

Reply:
Yes.  The right side includes mappings for some fields and intermediate structs which might not be listed on the left side.  For example, `$.ignition.version` doesn't get a translation from `base`.